File: hash.c
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include "clib.h"

/* Mozna bude potreba neco jako lock */

static unsigned int HASH_SIZE = 100000; // 13107200;

struct record {
    uint64_t hash;
    void *best;
    int depth:6;
    unsigned int player:1;
    unsigned int used:1;
};

static struct record *table;

static int reset_count=0, get_count=0, find_count=0, add_count=0, overwritten_count=0;

void info_hash()
{
    printf("info hashtable - reset: %d, get: %d, find: %d, add: %d, ow: %d\n",
            reset_count, get_count, find_count, add_count, overwritten_count);
}

/** size in kilobytes, 0 means only reset (no resize) */
void reset_hash(int size)
{
    reset_count++;
    get_count = find_count = add_count = overwritten_count = 0;
    if (table) free(table);

    if (size > 0)
        HASH_SIZE = size * 1000 / sizeof(struct record);

    table = (struct record *) calloc(HASH_SIZE, sizeof(struct record));
    if (!table) exit(2);
    if (size > 0)
        printf("log Set transposition table size to %dMB (%u entries)\n",
                size, HASH_SIZE);
}

#define ix ((hash ^ player) % HASH_SIZE)

void *get_hash(uint64_t hash, int player)
{
    get_count++;

    return table[ix].best;
}

int find_hash(uint64_t hash, int depth, int player)
{
    find_count++;

    return table[ix].used
        && table[ix].depth >= depth
        && table[ix].player == player
        && table[ix].hash  == hash;
}

void add_hash(uint64_t hash, int depth, int player, void *best)
{
    add_count++;

    overwritten_count += table[ix].used;
    table[ix] = (struct record) {hash, best, depth, player, 1};
}

int get_hash_size()
{
    return HASH_SIZE;
}

void *get_by_index(int i)
{
    return table[i].best;
}

int empty_by_index(int i)
{
    return table && table[i].best;
}

File: Hash.hs
{-# LANGUAGE ForeignFunctionInterface #-}
module Hash (
    infoHash,
    resetHash,
    findHash,
    getHash,
    addHash
) where

import Data.Int (Int64)
import Foreign.StablePtr
import BitRepresenation (DMove, Player, playerToInt)
import Control.Monad

foreign import ccall "clib.h info_hash"   infoHash :: IO ()
foreign import ccall "clib.h reset_hash" c_resetHash :: Int -> IO ()
foreign import ccall "clib.h find_hash" c_findHash :: Int64 -> Int -> Int -> IO Bool
foreign import ccall "clib.h get_hash"  c_getHash  :: Int64 -> Int -> IO (StablePtr (DMove, Int))
foreign import ccall "clib.h add_hash"  c_addHash  :: Int64 -> Int -> Int -> StablePtr (DMove, Int) -> IO ()
foreign import ccall "clib.h get_hash_size" getHashSize :: IO Int
foreign import ccall "clib.h get_by_index" getByIndex :: Int -> IO (StablePtr (DMove, Int))
foreign import ccall "clib.h empty_by_index" emptyByIndex :: Int -> IO Bool


addHash :: Int64 -- ^ hash
        -> Int   -- ^ depth
        -> Player -- ^ active player
        -> (DMove, Int) -- ^ (PV, value)
        -> IO ()
addHash h d pl best = do
        let plNum = playerToInt pl
        isNotEmpty <- c_findHash h 0 plNum
        when isNotEmpty (c_getHash h plNum >>= freeStablePtr)

        ptr <- newStablePtr best
        c_addHash h d plNum ptr
        return ()

getHash :: Int64 -> Player -> IO (DMove, Int)
getHash h pl = do
        let plNum = playerToInt pl
        bestPtr <- c_getHash h plNum
        deRefStablePtr bestPtr

findHash :: Int64 -> Int -> Player -> IO Bool
findHash h d pl = c_findHash h d plNum
        where plNum = playerToInt pl

resetHash :: Int -> IO ()
resetHash size = do
    size' <- getHashSize
    forM_ [0 .. size'-1] (\i -> do
        isNotEmpty <- emptyByIndex i
        when isNotEmpty (getByIndex i >>= freeStablePtr))
    c_resetHash size
